onnxscript/onnx_types.py

```diff
@@ -3,85 +3,107 @@
 # Licensed under the MIT License.
 # --------------------------------------------------------------------------
 
+from typing import Optional, Tuple, Union
+
 import onnx
 import onnx.helper
 
+# Representations of ONNX types in ONNX Script.
+# Currently restricted to tensor types.
+# Example type annotations in ONNX Script.
+#    x : FLOAT (a scalar-tensor of rank 0)
+#    x : FLOAT[...] (a tensor of unknown rank)
+#    x : FLOAT['M', 'N'] (a tensor of rank 2 of unknown dimensions, with symbolic names)
+#    x : FLOAT[128, 1024] (a tensor of rank 2 of known dimensions)
 
-class TensorType:
-    # Reference implementation placeholder
-    # represents a generic ONNX tensor type
-    def __init__(self, dtype=onnx.TensorProto.UNDEFINED, shape=None) -> None:
-        self.dtype = dtype
-        self.shape = shape
+DimType = Union[int, str, type(None)]
 
-    def __str__(self) -> str:
-        shapestr = str(self.shape) if self.shape else "[...]"
-        return onnx.TensorProto.DataType.Name(self.dtype) + shapestr
 
-    def __repr__(self) -> str:
-        return f"{self.__class__.__name__}(dtype={self.dtype!r}, shape={self.shape!r})"
+def check_dim(dim):
+    if not isinstance(dim, (int, str, type(None))):
```

Comment:
Use DimType?

Reply:
Not sure how. DimType is a type-alias, used for shortening type-annotations/signature. isinstance does not support use of type-alias with generics in it.

Reply:
SG!